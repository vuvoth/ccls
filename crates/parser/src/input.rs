use std::ops::Range;

use logos::Lexer;

use serde::Serialize;

use crate::token_kind::TokenKind;

#[derive(Debug, PartialEq, Serialize)]
pub struct Input<'a> {
    kind: Vec<TokenKind>,
    source: &'a str,
    position: Vec<Range<usize>>,
}

impl<'a> Input<'a> {
    pub fn new(source: &'a str) -> Self {
        let mut input = Input {
            source,
            kind: Vec::new(),
            position: Vec::new(),
        };

        let mut lex = Lexer::<TokenKind>::new(source);

        while let Some(tk) = lex.next() {
            if tk == TokenKind::CommentBlockOpen {
                let mut closed = false;
                let mut join_span = lex.span();
                while let Some(t) = lex.next() {
                    join_span.end = lex.span().end;
                    if t == TokenKind::CommentBlockClose {
                        closed = true;
                        break;
                    }
                }

                if closed {
                    input.kind.push(TokenKind::BlockComment);
                } else {
                    input.kind.push(TokenKind::Error);
                }
                input.position.push(join_span);
            } else {
                input.kind.push(tk);
                input.position.push(lex.span());
            }
        }

        input
    }

    pub fn token_value(&self, index: usize) -> Option<&'a str> {
        if index < self.kind.len() {
            Some(&self.source[self.position[index].start..self.position[index].end])
        } else {
            // return None for out of bound index
            None
        }
    }

    pub fn kind_of(&self, index: usize) -> TokenKind {
        if index < self.kind.len() {
            self.kind[index]
        } else {
            TokenKind::EOF
        }
    }

    pub fn position_of(&self, index: usize) -> Option<Range<usize>> {
        if index < self.kind.len() {
            Some(self.position[index].clone())
        } else {
            // return error for out of bound index
            None
        }
    }

    pub fn size(&self) -> usize {
        self.kind.len()
    }
}

#[cfg(test)]
mod tests {
    use super::Input;

    fn test(source: &str, snapshot_name: &str) {
        let input = Input::new(&source);

        insta::assert_yaml_snapshot!(snapshot_name, input);
    }

    #[test]
    fn test_comment_block() {
        let source = r#"
        /*a + b == 10*/
        a + 10
    "#;
        test(source, "test_comment_block");
    }

    #[test]
    fn test_comment_error() {
        let source = r#"
        pragma 2.1.1;
        /*a + b == 10*
        a + 10
        template

        /*
    "#;
        test(source, "test_comment_error");
    }

    #[test]
    fn test_pragma() {
        let source = r#"
        /* test pragma token kinds */

    pragma circom 2.0.0;

    "#;
        test(source, "test_pragma");
    }

    #[test]
    fn test_function() {
        let source = r#"
    function nbits(a) {
        var n = 1;
        var r = 0;
        while (n-1<a) {
            r++;
            n *= 2;
        }
        return r;
    }"#;
        test(source, "test_function");
    }

    #[test]
    fn test_operators() {
        let source = r#"
        ({[]})
        ;.,:
        && &
        || |
        != !
        === == =
        --> ==>
        <-- <==
        <= <
        >= >
        ++ += +
        -- -= -
        **= **
        * *=
        / /=
        \ \=
        % %=
        ^ ^=
        ~ ~=
        >> >>=
        << <<=
        & &=
        | |=
    }"#;
        test(source, "test_operators");
    }
}
