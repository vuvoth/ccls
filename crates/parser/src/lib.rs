pub mod event;
pub mod grammar;
pub mod parser;
pub mod token_kind;

pub use logos::Lexer;

pub mod input;
pub mod output;
