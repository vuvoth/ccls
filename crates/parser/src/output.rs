use crate::{event::Event, token_kind::TokenKind};

#[derive(Debug)]
pub enum Child {
    Token(usize), // position of token,
    Error(String),
    Tree(Tree),
}

#[derive(Debug)]
pub struct Tree {
    kind: TokenKind,
    children: Vec<Child>,
}

pub type Output = Tree;

impl Output {
    fn empty() -> Self {
        Tree {
            kind: TokenKind::ParserError,
            children: Vec::new(),
        }
    }

    pub fn kind(&self) -> TokenKind {
        self.kind
    }

    pub fn children(&self) -> &Vec<Child> {
        &self.children
    }
}

impl From<Vec<Event>> for Output {
    fn from(events: Vec<Event>) -> Self {
        let mut stack = Vec::new();
        if let Some((last, elements)) = events.split_last() {
            if !matches!(*last, Event::Close) {
                return Output::empty();
            }
            for event in elements {
                match event {
                    Event::Open { kind } => {
                        stack.push(Tree {
                            kind: *kind,
                            children: Vec::new(),
                        });
                    }
                    Event::Close => {
                        let tree = stack.pop().unwrap();

                        stack.last_mut().unwrap().children.push(Child::Tree(tree));
                    }
                    Event::TokenPosition(token) => {
                        stack
                            .last_mut()
                            .unwrap()
                            .children
                            .push(Child::Token(*token));
                    }
                    Event::ErrorReport(error) => {
                        stack
                            .last_mut()
                            .unwrap()
                            .children
                            .push(Child::Error(error.clone()));
                    }
                }
            }
        }
        // TODO: Make it more safe
        stack.pop().unwrap()
    }
}
