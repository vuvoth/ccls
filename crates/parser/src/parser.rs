use std::cell::Cell;

use crate::{
    event::Event, grammar::entry::Scope, input::Input, output::Output, token_kind::TokenKind,
};

pub struct Context {
    pub r_curly_count: i32,
}

pub struct Parser<'a> {
    pub(crate) input: &'a Input<'a>,
    pub context: Context,
    pos: usize,
    fuel: Cell<u32>,
    pub(crate) events: Vec<Event>,
}

#[derive(Clone, Copy, Debug)]
pub enum Marker {
    Open(usize),
    Close(usize),
}

#[derive(Debug)]
pub enum ParserError {
    InvalidEvents,
}

impl<'a> Parser<'a> {
    pub fn wrap_trivial_tokens(&mut self) -> TokenKind {
        loop {
            let kind = self.input.kind_of(self.pos);

            if kind.is_trivial() == false {
                return kind;
            }

            self.events.push(Event::Open { kind });

            self.fuel.set(256);
            self.events.push(Event::TokenPosition(self.pos));
            self.skip();

            self.events.push(Event::Close);
        }
    }

    pub fn open(&mut self) -> Marker {
        if self.events.len() > 0 {
            self.wrap_trivial_tokens();
        }

        let marker = Marker::Open(self.events.len());
        self.events.push(Event::Open {
            kind: TokenKind::Error,
        });
        marker
    }

    pub fn open_before(&mut self, marker_closed: Marker) -> Marker {
        match marker_closed {
            Marker::Close(index) => {
                let marker_opened = Marker::Open(index);
                self.events.insert(
                    index,
                    Event::Open {
                        kind: TokenKind::EOF,
                    },
                );
                marker_opened
            }
            _ => unreachable!(),
        }
    }

    pub fn close(&mut self, open_marker: Marker, kind: TokenKind) -> Marker {
        match open_marker {
            Marker::Open(index) => {
                self.events[index] = Event::Open { kind };
                self.events.push(Event::Close);
                Marker::Close(index)
            }
            _ => unreachable!(),
        }
    }

    pub fn advance(&mut self) {
        // assert!(!self.eof());
        self.fuel.set(256);
        let token = Event::TokenPosition(self.pos);
        self.events.push(token);
        self.skip();
    }

    pub fn advance_with_token(&mut self, index: usize) {
        // assert!(token.kind != TokenKind::EOF);
        if self.input.kind_of(index) != TokenKind::EOF {
            self.fuel.set(256);
            let token = Event::TokenPosition(index);
            self.events.push(token);
        }
    }

    pub fn advance_with_error(&mut self, _error: &str) {
        let m = self.open();
        // TODO: Error reporting.
        if !self.eof() {
            self.advance();
        }
        self.close(m, TokenKind::Error);
    }

    pub fn error_report(&mut self, error: String) {
        let m = self.open();

        let token = Event::ErrorReport(error);
        self.events.push(token);

        self.close(m, TokenKind::Error);
    }
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a Input) -> Self {
        Self {
            input,
            pos: 0,
            context: Context { r_curly_count: 0 },
            fuel: Cell::new(256),
            events: Vec::new(),
        }
    }

    pub fn inc_rcurly(&mut self) {
        self.context.r_curly_count += 1;
    }

    pub fn dec_rcurly(&mut self) {
        self.context.r_curly_count -= 1;
    }

    pub fn current(&mut self) -> TokenKind {
        self.wrap_trivial_tokens()
    }

    pub fn next(&mut self) -> TokenKind {
        if self.fuel.get() == 0 {
            panic!("parser is stuck");
        }
        self.fuel.set(self.fuel.get() - 1);
        if self.pos < self.input.size() {
            self.pos += 1;
            return self.input.kind_of(self.pos);
        }

        TokenKind::EOF
    }

    pub fn at(&mut self, kind: TokenKind) -> bool {
        self.current() == kind
    }

    pub fn at_any(&mut self, kinds: &[TokenKind]) -> bool {
        let current_kind = self.current();
        kinds.contains(&current_kind)
    }

    pub fn at_assign_token(&mut self) -> bool {
        let current_kind = self.current();
        current_kind.is_assign_token()
    }

    pub fn at_inline_assign_signal(&mut self) -> bool {
        let current_kind = self.current();
        current_kind.is_inline_assign_signal()
    }

    pub fn at_var_assign(&mut self) -> bool {
        let current_kind = self.current();
        current_kind.is_var_assign()
    }

    pub fn skip(&mut self) {
        self.next();
    }

    pub fn skip_if(&mut self, kinds: &[TokenKind]) {
        if self.at_any(kinds) {
            self.skip();
        }
    }

    pub fn eat(&mut self, kind: TokenKind) -> bool {
        if self.at(kind) {
            self.advance();
            return true;
        }

        false
    }

    pub fn expect_any(&mut self, kinds: &[TokenKind]) {
        let kind = self.current();
        if kinds.contains(&kind) {
            self.advance();
        } else {
            let error = format!("expect {:?} but got {:?}", kinds, kind);
            self.error_report(error);
        }
    }

    pub fn expect(&mut self, kind: TokenKind) {
        if self.at(kind) {
            self.advance();
        } else {
            let error = format!("expect {:?} but got {:?}", kind, self.current());
            self.error_report(error);
        }
    }

    pub fn eof(&mut self) -> bool {
        self.current() == TokenKind::EOF
    }
}

impl Parser<'_> {
    pub fn parsing_with_scope(input: &Input, scope: Scope) -> Output {
        let mut p = Parser::new(input);
        scope.parse(&mut p);
        Output::from(p.events)
    }

    pub fn parsing(input: &Input) -> Output {
        let c = Scope::CircomProgram;
        Parser::parsing_with_scope(input, c)
    }
}
