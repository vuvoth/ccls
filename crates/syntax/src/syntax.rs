use parser::grammar::entry::Scope;
use parser::input::Input;
use parser::output::{Child, Output};
use parser::parser::Parser;
use parser::token_kind::TokenKind;
use rowan::{GreenNode, GreenNodeBuilder};

pub use rowan::{
    api::Preorder, Direction, NodeOrToken, SyntaxText, TextRange, TextSize, TokenAtOffset,
    WalkEvent,
};

use crate::syntax_node::SyntaxNode;

pub struct SyntaxTreeBuilder<'a> {
    builder: GreenNodeBuilder<'static>,
    input: &'a Input<'a>,
}

impl<'a> SyntaxTreeBuilder<'a> {
    pub fn new(input: &'a Input) -> Self {
        Self {
            builder: GreenNodeBuilder::new(),
            input,
        }
    }
    pub fn build_rec(&mut self, tree: &Output) {
        self.builder.start_node(tree.kind().into());
        for child in tree.children() {
            match child {
                Child::Token(token_id) => {
                    let token_kind = self.input.kind_of(*token_id);
                    // TODO: return Error to replace .unwrap()
                    let token_value = self.input.token_value(*token_id).unwrap();
                    self.builder.start_node(token_kind.into());
                    self.builder.token(token_kind.into(), token_value);
                    self.builder.finish_node();
                }
                Child::Tree(child_tree) => self.build_rec(child_tree),
                Child::Error(error) => {
                    let token_kind = TokenKind::Error;
                    let token_value = error.as_str();

                    self.builder.start_node(token_kind.into());
                    self.builder.token(token_kind.into(), token_value);
                    self.builder.finish_node();
                }
            }
        }

        self.builder.finish_node();
    }

    pub fn build(&mut self, tree: Output) {
        self.build_rec(&tree);
    }

    pub fn finish(self) -> GreenNode {
        self.builder.finish()
    }

    pub fn syntax_tree(source: &str) -> SyntaxNode {
        let input = Input::new(source);

        let output = Parser::parsing(&input);

        let mut builder = SyntaxTreeBuilder::new(&input);
        builder.build(output);
        let green = builder.finish();
        SyntaxNode::new_root(green)
    }
}

pub fn syntax_node_from_source(source: &str, scope: Scope) -> SyntaxNode {
    let input = Input::new(&source);
    let output = Parser::parsing_with_scope(&input, scope);

    // output is a tree whose node is index of token, no content of token
    // convert output into green node
    let mut builder = SyntaxTreeBuilder::new(&input);
    builder.build(output);
    let green = builder.finish();

    // then cast green node into syntax node
    let syntax = SyntaxNode::new_root(green);

    syntax
}

#[cfg(test)]
mod tests {
    use crate::test_syntax;
    use parser::grammar::entry::Scope;

    #[test]
    fn pragma_happy_test() {
        test_syntax!("/src/test_files/happy/pragma.circom", Scope::Pragma);
    }

    #[test]
    fn template_happy_test() {
        // SOURCE & EXPECTED RESULT
        test_syntax!("/src/test_files/happy/template.circom", Scope::Template);
    }

    #[test]
    fn block_happy_test() {
        test_syntax!("/src/test_files/happy/block.circom", Scope::Block);
    }

    #[test]
    fn comment_happy_test() {
        test_syntax!(
            "/src/test_files/happy/block_comment.circom",
            Scope::CircomProgram
        );
        test_syntax!(
            "/src/test_files/happy/line_comment.circom",
            Scope::CircomProgram
        );
    }

    #[test]
    fn full_circom_program() {
        test_syntax!(
            "/src/test_files/happy/full_circom_program.circom",
            Scope::CircomProgram
        );
    }
}
