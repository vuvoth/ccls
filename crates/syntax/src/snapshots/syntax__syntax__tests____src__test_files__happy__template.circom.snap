---
source: crates/syntax/src/syntax.rs
expression: "crate :: view_syntax :: view_ast(& syntax)"
---
 TemplateDef 0..349
|      TemplateKw 0..8
|     |      TemplateKw 0..8 "template"
|      WhiteSpace 8..9
|     |      WhiteSpace 8..9 " "
|      TemplateName 9..20
|     |      Identifier 9..20
|     |     |      Identifier 9..20 "MultiplierN"
|      WhiteSpace 20..21
|     |      WhiteSpace 20..21 " "
|      ParameterList 21..31
|     |      LParen 21..22
|     |     |      LParen 21..22 "("
|     |      Identifier 22..23
|     |     |      Identifier 22..23 "N"
|     |      Comma 23..24
|     |     |      Comma 23..24 ","
|     |      WhiteSpace 24..25
|     |     |      WhiteSpace 24..25 " "
|     |      Identifier 25..26
|     |     |      Identifier 25..26 "P"
|     |      Comma 26..27
|     |     |      Comma 26..27 ","
|     |      WhiteSpace 27..28
|     |     |      WhiteSpace 27..28 " "
|     |      Identifier 28..30
|     |     |      Identifier 28..30 "QQ"
|     |      RParen 30..31
|     |     |      RParen 30..31 ")"
|      WhiteSpace 31..32
|     |      WhiteSpace 31..32 " "
|      Block 32..349
|     |      LCurly 32..33
|     |     |      LCurly 32..33 "{"
|     |      EndLine 33..35
|     |     |      EndLine 33..35 "\r\n"
|     |      WhiteSpace 35..47
|     |     |      WhiteSpace 35..47 "            "
|     |      CommentLine 47..87
|     |     |      CommentLine 47..87 "//Declaration of signals and components."
|     |      EndLine 87..89
|     |     |      EndLine 87..89 "\r\n"
|     |      WhiteSpace 89..101
|     |     |      WhiteSpace 89..101 "            "
|     |      StatementList 101..348
|     |     |      InputSignalDecl 101..119
|     |     |     |      SignalHeader 101..114
|     |     |     |     |      SignalKw 101..107
|     |     |     |     |     |      SignalKw 101..107 "signal"
|     |     |     |     |      WhiteSpace 107..108
|     |     |     |     |     |      WhiteSpace 107..108 " "
|     |     |     |     |      InputKw 108..113
|     |     |     |     |     |      InputKw 108..113 "input"
|     |     |     |     |      WhiteSpace 113..114
|     |     |     |     |     |      WhiteSpace 113..114 " "
|     |     |     |      ComplexIdentifier 114..119
|     |     |     |     |      Identifier 114..116
|     |     |     |     |     |      Identifier 114..116 "in"
|     |     |     |     |      LBracket 116..117
|     |     |     |     |     |      LBracket 116..117 "["
|     |     |     |     |      Expression 117..118
|     |     |     |     |     |      ExpressionAtom 117..118
|     |     |     |     |     |     |      Identifier 117..118
|     |     |     |     |     |     |     |      Identifier 117..118 "N"
|     |     |     |     |      RBracket 118..119
|     |     |     |     |     |      RBracket 118..119 "]"
|     |     |      Semicolon 119..120
|     |     |     |      Semicolon 119..120 ";"
|     |     |      EndLine 120..122
|     |     |     |      EndLine 120..122 "\r\n"
|     |     |      WhiteSpace 122..134
|     |     |     |      WhiteSpace 122..134 "            "
|     |     |      OutputSignalDecl 134..151
|     |     |     |      SignalHeader 134..148
|     |     |     |     |      SignalKw 134..140
|     |     |     |     |     |      SignalKw 134..140 "signal"
|     |     |     |     |      WhiteSpace 140..141
|     |     |     |     |     |      WhiteSpace 140..141 " "
|     |     |     |     |      OutputKw 141..147
|     |     |     |     |     |      OutputKw 141..147 "output"
|     |     |     |     |      WhiteSpace 147..148
|     |     |     |     |     |      WhiteSpace 147..148 " "
|     |     |     |      ComplexIdentifier 148..151
|     |     |     |     |      Identifier 148..151
|     |     |     |     |     |      Identifier 148..151 "out"
|     |     |      Semicolon 151..152
|     |     |     |      Semicolon 151..152 ";"
|     |     |      EndLine 152..154
|     |     |     |      EndLine 152..154 "\r\n"
|     |     |      WhiteSpace 154..166
|     |     |     |      WhiteSpace 154..166 "            "
|     |     |      ComponentDecl 166..185
|     |     |     |      ComponentKw 166..175
|     |     |     |     |      ComponentKw 166..175 "component"
|     |     |     |      WhiteSpace 175..176
|     |     |     |     |      WhiteSpace 175..176 " "
|     |     |     |      ComplexIdentifier 176..185
|     |     |     |     |      Identifier 176..180
|     |     |     |     |     |      Identifier 176..180 "comp"
|     |     |     |     |      LBracket 180..181
|     |     |     |     |     |      LBracket 180..181 "["
|     |     |     |     |      Expression 181..184
|     |     |     |     |     |      Sub 181..184
|     |     |     |     |     |     |      ExpressionAtom 181..182
|     |     |     |     |     |     |     |      Identifier 181..182
|     |     |     |     |     |     |     |     |      Identifier 181..182 "N"
|     |     |     |     |     |     |      Sub 182..183
|     |     |     |     |     |     |     |      Sub 182..183 "-"
|     |     |     |     |     |     |      ExpressionAtom 183..184
|     |     |     |     |     |     |     |      Number 183..184
|     |     |     |     |     |     |     |     |      Number 183..184 "1"
|     |     |     |     |      RBracket 184..185
|     |     |     |     |     |      RBracket 184..185 "]"
|     |     |      Semicolon 185..186
|     |     |     |      Semicolon 185..186 ";"
|     |     |      EndLine 186..188
|     |     |     |      EndLine 186..188 "\r\n"
|     |     |      WhiteSpace 188..201
|     |     |     |      WhiteSpace 188..201 "             "
|     |     |      EndLine 201..203
|     |     |     |      EndLine 201..203 "\r\n"
|     |     |      WhiteSpace 203..215
|     |     |     |      WhiteSpace 203..215 "            "
|     |     |      CommentLine 215..228
|     |     |     |      CommentLine 215..228 "//Statements."
|     |     |      EndLine 228..230
|     |     |     |      EndLine 228..230 "\r\n"
|     |     |      WhiteSpace 230..242
|     |     |     |      WhiteSpace 230..242 "            "
|     |     |      ForLoop 242..328
|     |     |     |      ForKw 242..245
|     |     |     |     |      ForKw 242..245 "for"
|     |     |     |      LParen 245..246
|     |     |     |     |      LParen 245..246 "("
|     |     |     |      VarDecl 246..255
|     |     |     |     |      VarKw 246..249
|     |     |     |     |     |      VarKw 246..249 "var"
|     |     |     |     |      WhiteSpace 249..250
|     |     |     |     |     |      WhiteSpace 249..250 " "
|     |     |     |     |      ComplexIdentifier 250..252
|     |     |     |     |     |      Identifier 250..251
|     |     |     |     |     |     |      Identifier 250..251 "i"
|     |     |     |     |     |      WhiteSpace 251..252
|     |     |     |     |     |     |      WhiteSpace 251..252 " "
|     |     |     |     |      Assign 252..253
|     |     |     |     |     |      Assign 252..253 "="
|     |     |     |     |      WhiteSpace 253..254
|     |     |     |     |     |      WhiteSpace 253..254 " "
|     |     |     |     |      Expression 254..255
|     |     |     |     |     |      ExpressionAtom 254..255
|     |     |     |     |     |     |      Number 254..255
|     |     |     |     |     |     |     |      Number 254..255 "0"
|     |     |     |      Semicolon 255..256
|     |     |     |     |      Semicolon 255..256 ";"
|     |     |     |      WhiteSpace 256..257
|     |     |     |     |      WhiteSpace 256..257 " "
|     |     |     |      Expression 257..264
|     |     |     |     |      LessThan 257..264
|     |     |     |     |     |      ExpressionAtom 257..258
|     |     |     |     |     |     |      Identifier 257..258
|     |     |     |     |     |     |     |      Identifier 257..258 "i"
|     |     |     |     |     |      WhiteSpace 258..259
|     |     |     |     |     |     |      WhiteSpace 258..259 " "
|     |     |     |     |     |      LessThan 259..260
|     |     |     |     |     |     |      LessThan 259..260 "<"
|     |     |     |     |     |      WhiteSpace 260..261
|     |     |     |     |     |     |      WhiteSpace 260..261 " "
|     |     |     |     |     |      Sub 261..264
|     |     |     |     |     |     |      ExpressionAtom 261..262
|     |     |     |     |     |     |     |      Identifier 261..262
|     |     |     |     |     |     |     |     |      Identifier 261..262 "N"
|     |     |     |     |     |     |      Sub 262..263
|     |     |     |     |     |     |     |      Sub 262..263 "-"
|     |     |     |     |     |     |      ExpressionAtom 263..264
|     |     |     |     |     |     |     |      Number 263..264
|     |     |     |     |     |     |     |     |      Number 263..264 "1"
|     |     |     |      Semicolon 264..265
|     |     |     |     |      Semicolon 264..265 ";"
|     |     |     |      WhiteSpace 265..266
|     |     |     |     |      WhiteSpace 265..266 " "
|     |     |     |      AssignStatement 266..269
|     |     |     |     |      Expression 266..269
|     |     |     |     |     |      UnitInc 266..269
|     |     |     |     |     |     |      ExpressionAtom 266..267
|     |     |     |     |     |     |     |      Identifier 266..267
|     |     |     |     |     |     |     |     |      Identifier 266..267 "i"
|     |     |     |     |     |     |      UnitInc 267..269
|     |     |     |     |     |     |     |      UnitInc 267..269 "++"
|     |     |     |      RParen 269..270
|     |     |     |     |      RParen 269..270 ")"
|     |     |     |      Block 270..328
|     |     |     |     |      LCurly 270..271
|     |     |     |     |     |      LCurly 270..271 "{"
|     |     |     |     |      EndLine 271..273
|     |     |     |     |     |      EndLine 271..273 "\r\n"
|     |     |     |     |      WhiteSpace 273..289
|     |     |     |     |     |      WhiteSpace 273..289 "                "
|     |     |     |     |      StatementList 289..327
|     |     |     |     |     |      AssignStatement 289..312
|     |     |     |     |     |     |      Expression 289..297
|     |     |     |     |     |     |     |      ArrayQuery 289..296
|     |     |     |     |     |     |     |     |      ExpressionAtom 289..293
|     |     |     |     |     |     |     |     |     |      Identifier 289..293
|     |     |     |     |     |     |     |     |     |     |      Identifier 289..293 "comp"
|     |     |     |     |     |     |     |     |      LBracket 293..294
|     |     |     |     |     |     |     |     |     |      LBracket 293..294 "["
|     |     |     |     |     |     |     |     |      Expression 294..295
|     |     |     |     |     |     |     |     |     |      ExpressionAtom 294..295
|     |     |     |     |     |     |     |     |     |     |      Identifier 294..295
|     |     |     |     |     |     |     |     |     |     |     |      Identifier 294..295 "i"
|     |     |     |     |     |     |     |     |      RBracket 295..296
|     |     |     |     |     |     |     |     |     |      RBracket 295..296 "]"
|     |     |     |     |     |     |     |      WhiteSpace 296..297
|     |     |     |     |     |     |     |     |      WhiteSpace 296..297 " "
|     |     |     |     |     |     |      Assign 297..298
|     |     |     |     |     |     |     |      Assign 297..298 "="
|     |     |     |     |     |     |      WhiteSpace 298..299
|     |     |     |     |     |     |     |      WhiteSpace 298..299 " "
|     |     |     |     |     |     |      Expression 299..312
|     |     |     |     |     |     |     |      Call 299..312
|     |     |     |     |     |     |     |     |      ExpressionAtom 299..310
|     |     |     |     |     |     |     |     |     |      Identifier 299..310
|     |     |     |     |     |     |     |     |     |     |      Identifier 299..310 "Multiplier2"
|     |     |     |     |     |     |     |     |      LParen 310..311
|     |     |     |     |     |     |     |     |     |      LParen 310..311 "("
|     |     |     |     |     |     |     |     |      RParen 311..312
|     |     |     |     |     |     |     |     |     |      RParen 311..312 ")"
|     |     |     |     |     |      Semicolon 312..313
|     |     |     |     |     |     |      Semicolon 312..313 ";"
|     |     |     |     |     |      EndLine 313..315
|     |     |     |     |     |     |      EndLine 313..315 "\r\n"
|     |     |     |     |     |      WhiteSpace 315..327
|     |     |     |     |     |     |      WhiteSpace 315..327 "            "
|     |     |     |     |      RCurly 327..328
|     |     |     |     |     |      RCurly 327..328 "}"
|     |     |      EndLine 328..330
|     |     |     |      EndLine 328..330 "\r\n"
|     |     |      WhiteSpace 330..346
|     |     |     |      WhiteSpace 330..346 "                "
|     |     |      EndLine 346..348
|     |     |     |      EndLine 346..348 "\r\n"
|     |      RCurly 348..349
|     |     |      RCurly 348..349 "}"
